File: Downloader.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;

/**
 * This script was created by Shahid Karim on 7/15/19.
 */
public class Downloader {
    private static String scriptVersion = "0.4a";
    private static String downloadFile = "";
    private static String downloadLocation = "";
    private static int numberOfThreads = 1;

    private static String YOUTUBE_DL_LOCATION = "";
    private static String PHANTOM_JS_LOCATION = "";
    private static String ARIA_2C_LOCATION = "";

    private static void printRequiredArguments() {
        System.out.println("downloadFile - The location of the file that contains URLS.");
        System.out.println("downloadLocation - The location where the files should be downloaded.");
        System.out.println("numberOfThreads - The number of threads to use in order to speed up the download process.");
        System.out.println("\nExample usage: java Downloader.java C:/DL.txt" + " " + "C:/Video Downloads" + " 10");
    }

    private static void checkArguments(String[] args) {
        if(args.length > 0) {
            // Options - The usual command line options such as -h, --help for help and -v, --version for version.
            if (args[0].contentEquals("-h") || args[0].contentEquals("--help")) {
                System.out.println("\nThis script is a command line wrapper written in Java that uses youtube-dl and aria2c " +
                        "to download videos when given a file that contains URLs. The required command line arguments are as follows: \n");
                printRequiredArguments();
            } else if (args[0].contentEquals("-v") || args[0].contentEquals("--version")) {
                System.out.println("Version " + scriptVersion);
            }
        }

        // Edge cases where there are either no inputs (too few arguments) or 4 or more inputs (too many arguments)
        if(args.length <= 0) {
            System.out.println("Not enough arguments. This script requires the following: \n");
            printRequiredArguments();
        } else if (args.length > 3) {
            System.out.println("Too many arguments.\n");
            printRequiredArguments();
        }

        // Normal case where there is more than 0 inputs but less than the 3 required inputs:
        if(args.length > 0 && args.length < 3) {
            System.out.println("Not enough arguments. This script requires the following: \n");
            printRequiredArguments();
        }

        // Normal case where there is exactly 3 arguments.
        if (args.length == 3) {
            downloadFile = args[0];
            downloadLocation = args[1];
            try {
                numberOfThreads = Integer.parseInt(args[2]);
            } catch (NumberFormatException e) {
                System.out.println("An integer is required as the third argument: numberOfThreads." + " Defaulting to 1 thread.");
            }

            // Do not allow the user to set numberOfThreads higher than the threads available on their CPU.
            if(numberOfThreads > Runtime.getRuntime().availableProcessors()) {
                numberOfThreads = Runtime.getRuntime().availableProcessors();
                System.out.println("The numberOfThreads specified was above the number of threads your CPU supports. " +
                        "Defaulting to the max amount possible on your CPU: " + numberOfThreads);
            }

            System.out.println("Starting the download script with the following inputs: ");
            System.out.println(downloadFile);
            System.out.println(downloadLocation);
            System.out.println(numberOfThreads);
        }
    }

    public static void setup() {
        YOUTUBE_DL_LOCATION = System.getProperty("user.dir") + "\\Youtubedl\\youtube-dl.exe";
        YOUTUBE_DL_LOCATION = YOUTUBE_DL_LOCATION.replace("\\", "/");

        PHANTOM_JS_LOCATION = System.getProperty("user.dir") + "\\PhantomJS\\bin\\phantomjs.exe";
        PHANTOM_JS_LOCATION = PHANTOM_JS_LOCATION.replace("\\", "/");

        ARIA_2C_LOCATION = System.getProperty("user.dir") + "\\Aria2c\\aria2c.exe";
        ARIA_2C_LOCATION = ARIA_2C_LOCATION.replace("\\", "/");
    }

    private static String youtubedl(String command, String url) {
        StringBuilder standardOutputStringBuilder = new StringBuilder();
        StringBuilder errorStringBuilder = new StringBuilder();

        try {
            Runtime rt = Runtime.getRuntime();
            Process process = rt.exec("cmd /c " + "\"" + YOUTUBE_DL_LOCATION + "\"" + command + " " + url);
            InputStream stdin = process.getInputStream();
            InputStreamReader isr = new InputStreamReader(stdin);
            BufferedReader br = new BufferedReader(isr);

            InputStream stderr = process.getErrorStream();
            InputStreamReader esr = new InputStreamReader(stderr);
            BufferedReader errorReader = new BufferedReader(esr);

            String line = null;
            while ( (line = br.readLine()) != null) {
                standardOutputStringBuilder.append(line);
            }
            while ( (line = errorReader.readLine()) != null) {
                errorStringBuilder.append(line);
            }

            //final Process p = Runtime.getRuntime().exec("cmd /c " + "\"" + YOUTUBE_DL_LOCATION + "\"" +" --get-url" + " " + url);
        } catch (IOException e) {
            System.out.println("Could not find the executable.");
        }

        return standardOutputStringBuilder.toString();
    }

    private static void download(String unconvertedUrl) {
        String convertedUrl = youtubedl(" --get-url", unconvertedUrl);
        String name = youtubedl(" --get-filename -o '%(title)s.%(ext)s' --restrict-filenames", unconvertedUrl);
        System.out.println(unconvertedUrl);
        System.out.println(name.replace("\'", ""));
        System.out.println(convertedUrl);
    }

    public static void main(String[] args) {
        checkArguments(args);
        setup();
        download("https://www.youtube.com/watch?v=dY6jR52fFWo");
    }
}

File: downloader.py
# This script was created by Shahid Karim on 7/14/19
# This script requires that python3, youtube-dl, and aria2c have been added to your PATH.
# This script has only been tested on Windows.
# Run this script through the command line: 
# py downloader.py downloadFile downloadLocation numberOfThreads
# py downloader.py "DL.txt" "C:/Transfer Files/Videos" 7

import subprocess
import argparse
from multiprocessing import Pool

downloadFile = ""
downloadLocation = ""
numberOfThreads = 5   
text = 'This script will use youtube-dl and aria2c to download any media when given a file with the URL(s) to the media you want to download. Example usage: py downloader.py "DL.txt" "C:/Transfer Files/Videos" 7'

# Stores the URLS from a given file into an array.
def load_urls(fileName):
    print("Loading URLS into an array.")
    fileData = []
    with open(fileName, 'r') as downloadFile:
        for line in downloadFile:
            fileData.append(line.rstrip())
    return fileData

# Uses youtube-dl to find the URL where the media is stored and gets the media's human-readable filename. Then it uses aria2c in order to download to the directory that was specified through the command line.
def download(unconvertedUrl):
    global downloadFile
    global downloadLocation
    global numberOfThreads

    # Get the URL of the media itself (mp4, jpg, png, gif, etc) rather than just the html page (index.html). 
    convertedUrl = subprocess.run(["youtube-dl", "--get-url", unconvertedUrl], stdout=subprocess.PIPE).stdout.decode('utf-8')

    # Get the actual name of the media file (butterfly.png) rather than the server's identifier name (some random string 2350wdf320423.png)
    filename = subprocess.run(["youtube-dl", "--get-filename", "-o '%(title)s.%(ext)s'", "--restrict-filenames", unconvertedUrl], stdout=subprocess.PIPE).stdout.decode('utf-8')
    filename = filename.replace("'", "")

    #print(downloadLocation + "/" + filename.strip())
    #location = downloadLocation + "/" + filename.strip()
    
    print("Starting download of " + filename.strip())
    # Make sure that the media url isn't equal to an empty string.
    if convertedUrl:
        # Make sure that the filename isn't equal to an empty string.
        if filename:
            # Start downloading the media file using aria2c
            # -x The maximum number of connections to one server for each download
            # -j Set the maximum number of parallel downloads for every queue item.
            # -s Download a file using the given number of connections
            # -o The file name of the downloaded file.
            # -c Will continue a download if it was stopped for any reason and will not redownload any files if it already exists.
            # --auto-file-renaming If a control file (.aria2) exists then the download will resume, otherwise if no control file exists then since --allow-overwrite is false by default the file will not be re-downloaded.
            subprocess.run(["aria2c", '--auto-file-renaming=false', '-c', '-x5', '-j7', '-s5', '-o', filename, '-d', downloadLocation, convertedUrl.strip()], stdout=subprocess.PIPE).stdout.decode('utf-8')
    
    print("Download of " + filename.strip() + " has finished.")
    
# Entry point for the program.    
def main(args):
    update_vars(args)
    
    # Store the URLs contained within the file given in the command line arguments into a local array variable.
    data = load_urls(args.downloadFile)
    
    # If this file is imported as a module then none of this will run.
    # But since this file will be run directly then the following lines will always run.
    # This line is required when doing anything with multiprocessing in order to avoid infinite recursive calls.
    if __name__ == "__main__":
        # Using multiprocessing, download all the media contained in the data array URLs. 
        with Pool(numberOfThreads) as p:
            p.map(download, data)

# Updates global variables based on command line arguments. 
def update_vars(args):
    global downloadFile
    global downloadLocation
    global numberOfThreads
    
    downloadFile = args.downloadFile
    downloadLocation = args.downloadLocation
    if(args.numberOfThreads <= 0):
        numberOfThreads = 1
    else:
        numberOfThreads = args.numberOfThreads
   
# The command line contains 4 arguments:
# 0. This script's name: downloader.py
# 1. File containing all the URLs to download: DL.txt
# 2. Download location: C:\Users
# 3. Number of threads to use for downloading: 10

parser = argparse.ArgumentParser(description = text)
parser.add_argument("downloadFile", help="The location of the file that contains links to your media downloads.")
parser.add_argument("downloadLocation", help="The location you want to store the downloaded media.")
parser.add_argument("numberOfThreads", help="The number threads do you want to use to speed up the download of your media.", type=int)

args = parser.parse_args()

main(args)